---
title: "STAR Protocols Quick Data Viz"
author: "Haylie Helms"
date: "2025-05-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
library(Seurat)
library(ggplot2)
library(tidyverse)
library(patchwork)
library(dplyr)
library(knitr)
library(DT)
library(htmltools)

setwd("E:/Haylie/VHD_scPrint1-HE/StarProtocols_Analysis")
```


# Visium HD Spatial Transcriptomics - Seurat Standards

## Load Data and identify which barcode binning size you want 2, 8, or 16
```{r}
data.dir = "E:/Haylie/VHD_scPrint1-HE/outs"

VHD1_data <- Load10X_Spatial(data.dir = data.dir, bin.size = c(8)) #c(2,8,16)) for all

# Setting default assay changing between 2 um, 8um, and 16um binning
# Assays(VHD1_data)
# DefaultAssay(VHD1_data) <- "Spatial.008um"
```

## OPTIONAL: Sequencing and Alignment summary from SpaceRanger
Note: Mapping and sequencing values should be %. 
For example: Valid UMIs = 99.9%
```{r}
# Load the metrics summary CSV file
SRsum <- read.csv("E:/Haylie/VHD_scPrint1-HE/outs/metrics_summary.csv", header = TRUE, check.names = FALSE)

# Transpose the data (convert columns to rows)
SRsum_t <- as.data.frame(t(SRsum))

# The second column has "V1" column name I want to delete - renaming it to an empty string
colnames(SRsum_t)[1] <- "Value"

# Display Table - static option
#kable(SRsum_t, caption = "SpaceRanger 'Metrics Summary'")

# Display table - interactive version
htmltools::tagList(
  h3("SpaceRanger Output Metric Summary"),  
  datatable(SRsum_t, options = list(pageLength = 10)))
```

# UMI Counts 
nCount_Spatial
```{r}
vln.plot1 <- VlnPlot(VHD1_data, features = "nCount_Spatial.008um", pt.size = 0.005) + theme(axis.text = element_text(size = 10), plot.title = element_text(size = 14, face = "bold", hjust = 0.5)) + ggtitle("Distribution of UMI counts per spot") + NoLegend()

# Set the file path and export
ggsave(
  filename = "UMI_violinplot_raw.png",
  plot = vln.plot1,
  width = 5, 
  height = 4,  
  dpi = 300,
  units = "in")
  
vln.plot1
```

```{r}
# Remove dots by making pt.size = 0 
vln.plot2 <- VlnPlot(VHD1_data, features = "nCount_Spatial.008um", pt.size = 0.00) + theme(axis.text = element_text(size = 10), plot.title = element_text(size = 14, face = "bold", hjust = 0.5)) + ggtitle("Distribution of UMI counts per spot") + NoLegend()

# Set the file path and export
ggsave(
  filename = "UMI_violinplot_raw_nospots.png",
  plot = vln.plot2,
  width = 5, 
  height = 4,  
  dpi = 300,
  units = "in")

vln.plot2
```

# UMI Heatmap
```{r}
count.plot1 <- SpatialFeaturePlot(VHD1_data, features = "nCount_Spatial.008um") + theme(legend.position = "right") + ggtitle("UMI Counts Spatial Distribution")

# Set the file path and export
ggsave(
  filename = "UMI_heatmap_raw.png",
  plot = count.plot1,
  width = 5, 
  height = 4,  
  dpi = 300,
  units = "in")

count.plot1
```

```{r}
# Adjust scale so we can see distribution

# Inspect data spread (or just look at your violin plot)
summary(VHD1_data$nCount_Spatial.008um)

Adj.count.plot <- SpatialFeaturePlot(VHD1_data, features = "nCount_Spatial.008um", min.cutoff = 0, max.cutoff = 60) + theme(legend.position = "right")

# Set the file path and export
ggsave(
  filename = "UMI_heatmap_adjusted.png",
  plot = Adj.count.plot,
  width = 5, 
  height = 4,  
  dpi = 300,
  units = "in")

Adj.count.plot
```

# Label Print Patterns/Samples/ROIs
Print pattern (or sample or region of interest) information is generated by manual annotation in Loupe Browser via the selection tool (freehand, rectangle, or paintbrush) and "Custom Groups" --> Download barcode groupings as .csv --> included unlabeled.
```{r}
# Add print pattern to the meta data
printpattern_df <- read.csv("E:/Haylie/VHD_scPrint1-HE/StarProtocols_Analysis/HE-PrintPattern-Updated.csv")

# If there isn't an assigned print pattern add "Unlabeled"
printpattern_df$PrintPattern[printpattern_df$PrintPattern == "" | is.na(printpattern_df$PrintPattern)] <- "Unlabeled"

unique(printpattern_df$PrintPattern)

head(printpattern_df)

# Make sure that the barcode is the rowname
rownames(printpattern_df) <- printpattern_df$Barcode

# Get rid of the duplicate barcode
printpattern_df$Barcode <- NULL

head(printpattern_df)

# Assign to meta data
VHD1_data <- AddMetaData(VHD1_data, metadata = printpattern_df)

head(VHD1_data)

# Confirm no missing annotations
sum(is.na(VHD1_data@meta.data$PrintPattern))

# Visualize
Haylie_colors19 <- c("#ff0000", "#f44336","#e81e63",  "#9c27b0", "#673ab7", "#3f51b5","#2196f3", "#03a9f4","#00bcd4", "#009688", "#4caf50", "#8bc34a", "#cddc39", "#ffeb3b", "#ffc107", "#ff9800", "#ffffff", "#b1a4b5", "#000000")

printpattern.spatial.plot <- SpatialDimPlot(VHD1_data, group.by = "PrintPattern", pt.size.factor = 5) + scale_fill_manual(values = Haylie_colors19) + theme(legend.position = "right") + guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) + labs(fill = "Print Pattern") + ggtitle("Printed Populations") + theme(plot.title = element_text(size = 18, face = "bold", hjust = 0.5))

# Set the file path and export
ggsave(
  filename = "PrintPatterns_spatialmap.png",
  plot = printpattern.spatial.plot,
  width = 6, 
  height = 5,  
  dpi = 300,
  units = "in")

# Export with rasterized geoms
library(ggrastr)
ggsave("PrintPattern_HD.png", plot = rasterise(printpattern.spatial.plot, layers = 1), 
       width = 8, height = 5, dpi = 300)

printpattern.spatial.plot
```

# Label Cell Types
Cell type annotations are generated via another custom grouping in Loupe browser. My sample has fluorescent tags in it so that the nucleus or cell body has a unique color. Since Loupe (and Space Ranger to generate a Loupe file) do not allow multiple images to be loaded in at a time, I use our custom overlay software (https://zenodo.org/records/10511395) which has adjustable opacity so I can look at both images at the same time if using the H&E Space Ranger outputs. By toggling between the images I can manually annotate each cell using the paintbrush tool in Loupe.
```{r}
# For illustrative purposes I'm sharing an example of quick cell type annotations I did on the Loupe file made from the Space Ranger run with the IF image, eliminating the need for the overlay software. It is the same experiment/FASTQs, but I learned H&E and IF image Space Ranger outputs gave different barcode IDs. Our upcoming manuscript describing our single-cell bioprinting method, which uses this STAR protocol to perform the ST, is annotated via the overlay software. Feel free to contact me if you have questions.

# Load the CSV, ensuring the Barcode column becomes row names
Phenotype_df <- read.csv("E:/Haylie/VHD_scPrint1-HE/StarProtocols_Analysis/CellType.csv", row.names = 1)

# If there isn't an assigned cell type add "Unlabeled"
Phenotype_df$CellType[Phenotype_df$CellType == "" | is.na(Phenotype_df$CellType)] <- "Unlabeled"

unique(Phenotype_df$CellType)

head(Phenotype_df)

# Have to switch to the IF version of my data set so the barcodes align.
data.dir2 = "E:/Haylie/VHD_scPrint1/A1_scPrint1-spacerangerOUTS/outs"

VHD1_data_IF <- Load10X_Spatial(data.dir = data.dir2, bin.size = c(8)) #c(2,8,16)) for all

# Assign cell type to meta data
VHD1_data_IF <- AddMetaData(VHD1_data_IF, metadata = Phenotype_df)

head(VHD1_data_IF)

# Confirm no missing annotations
sum(is.na(VHD1_data_IF@meta.data$CellType))

Haylie_colors8 <- c("#ff69b4","#ffba00", "#c7ef00", "#00d7ff", "#deaaff","#ffff00", "#ff0000","#b1a4b5")

celltype.spatial.plot <- SpatialDimPlot(VHD1_data_IF, group.by = "CellType") + scale_fill_manual(values = Haylie_colors8) + theme(legend.position = "right") + guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) + labs(fill = "Cell Type") + ggtitle("Cell Type") + theme(plot.title = element_text(size = 18, face = "bold", hjust = 0.5))

# Set the file path and export
ggsave(
  filename = "CellType_spatialmap.png",
  plot = celltype.spatial.plot,
  width = 6, 
  height = 5,  
  dpi = 300,
  units = "in")

celltype.spatial.plot
```

```{r}
# Crop to an area that is fully annotated - Random TME

# Add print pattern from the IF data set to the meta data
IF_printpattern_df <- read.csv("E:/Haylie/VHD_scPrint1-HE/StarProtocols_Analysis/IF-PrintPattern.csv")

# If there isn't an assigned print pattern add "Unlabeled"
IF_printpattern_df$PrintPattern[IF_printpattern_df$PrintPattern == "" | is.na(IF_printpattern_df$PrintPattern)] <- "Unlabeled"

head(IF_printpattern_df)

# Make the barcode the rowname
rownames(IF_printpattern_df) <- IF_printpattern_df$Barcode

# Get rid of the duplicate barcode
IF_printpattern_df$Barcode <- NULL

head(IF_printpattern_df)

# Add to meta data
VHD1_data_IF <- AddMetaData(VHD1_data_IF, metadata = IF_printpattern_df)

head(VHD1_data_IF)

# Confirm no missing annotations
sum(is.na(VHD1_data@meta.data$PrintPattern))

# Subset just Random
Ran_Subset_VHD1_IF <- subset(VHD1_data_IF, subset = PrintPattern == "RandomPrint_All")

Haylie_colors3 <- c("#FF40FF",  "#FF2600", "#11ff00")

cropped_celltype.spatial.plot <- SpatialDimPlot(Ran_Subset_VHD1_IF, group.by = "CellType", pt.size.factor = 5) + scale_fill_manual(values = Haylie_colors3) + theme(legend.position = "right") + guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) + labs(fill = "Cell Type") + ggtitle("Cell Type") + theme(plot.title = element_text(size = 18, face = "bold", hjust = 0.5))

# Export with rasterized geoms
library(ggrastr)
ggsave("CellType_HD.png", plot = rasterise(cropped_celltype.spatial.plot, layers = 1), 
       width = 6, height = 3, dpi = 300)

# Set the file path and export
ggsave(
  filename = "NC_CellType_spatialmap.png",
  plot = cropped_celltype.spatial.plot,
  width = 6, 
  height = 5,  
  dpi = 300,
  units = "in")

cropped_celltype.spatial.plot
```

```{r}
# Showing where that cropped area came from
Haylie_colors19x <- c("#000000", "#000000","#000000",  "#000000", "#000000", "#000000","#000000", "#000000","#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#ffffff", "#000000", "#000000")

# Adjust the H&E print pattern ROI to align with the IF print pattern ROI (different space ranger pipelines and barcodes can't just overlay)
Adjusted_HE_printpattern_df <- read.csv("E:/Haylie/VHD_scPrint1-HE/StarProtocols_Analysis/Adjusted-HE-PrintPattern.csv")

# If there isn't an assigned print pattern add "Unlabeled"
Adjusted_HE_printpattern_df$PrintPattern[Adjusted_HE_printpattern_df$PrintPattern == "" | is.na(Adjusted_HE_printpattern_df$PrintPattern)] <- "Unlabeled"

unique(Adjusted_HE_printpattern_df$PrintPattern)

# Make sure that the barcode is the rowname
rownames(Adjusted_HE_printpattern_df) <- Adjusted_HE_printpattern_df$Barcode

# Get rid of the duplicate barcode
Adjusted_HE_printpattern_df$Barcode <- NULL

# Duplicating the Seurat object to not copy over it
Adjusted_VHD1_data <- VHD1_data

# Adding adjusted ROI to meta data 
Adjusted_VHD1_data <- AddMetaData(Adjusted_VHD1_data, metadata = Adjusted_HE_printpattern_df)

ROI.printpattern.spatial.plot <- SpatialDimPlot(Adjusted_VHD1_data, group.by = "PrintPattern") + scale_fill_manual(values = Haylie_colors19x) + theme(legend.position = "right") + guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) + labs(fill = "Print Pattern") + ggtitle("Printed Populations") + theme(plot.title = element_text(size = 18, face = "bold", hjust = 0.5))

# Set the file path and export
ggsave(
  filename = "ROI_PrintPatterns_spatialmap.png",
  plot = ROI.printpattern.spatial.plot,
  width = 6, 
  height = 5,  
  dpi = 300,
  units = "in")

# Export with rasterized geoms
library(ggrastr)
ggsave("ROIPrintPattern_HD.png", plot = rasterise(ROI.printpattern.spatial.plot, layers = 1), 
       width = 8, height = 5, dpi = 300)

ROI.printpattern.spatial.plot
```
